model_utils/tracker.py

```diff
@@ -86,11 +86,46 @@ def __delete__(self, obj):
         self.descriptor.__delete__(obj)
 
 
+class FieldsContext:
+    def __init__(self, tracker, *fields, state=None):
+        if state is None:
+            state = {}
```

Comment:
Or maybe a `defaultdict(0)` ?

Reply:
`state` may be passed as a parameter to `FieldsContext.__init__`, it will be not so obvious that there should be `defaultdict` instead of `dict`.

Reply:
Hi @tumb1er I'm not really sure this will reduce the complexity, so I guess you'd better leave it as it is. Maybe you could also here add a docstring that explains the architecture.

Reply:
Complexity - bad, docstrings - good. I'll do it.

Reply:
is it possible to split this pr in smaller ones?

Reply:
In this PR:
* Two classes for new mechanics were introduced for fields state tracking
* Some helper methods added that simplify using new mechanics in new code
* default monkey-patched methods refactored to use new mechanics

I can split it, but each one will be more or less useless without another one.